### PR TITLE
Carriers - CQRS - Add/Update Shipping locations and costs

### LINKED ISSUE #35927: [CQRS] Add/Update Shipping locations and costs (Free shipping, Tax...)
Update CQRS command (create, update) and query (get)
Add these fields:
- handling costs
- free shipping
- billing
- tax
- out of range behavior

Add related behat tests

### PR DESCRIPTION
**Pull request by boherm**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add/Update Shipping locations and costs
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | Only CI for now
| UI Tests          | Only CI for now
| Fixed issue or discussion?     | Fixes #35927 
| Related PRs       | 
| Sponsor company   | PrestaShop SA

### CODE REVIEW

File: src/Adapter/Carrier/CommandHandler/AddCarrierHandler.php
Comment (jolelievre) [inline src/Adapter/Carrier/CommandHandler/AddCarrierHandler.php]: I'm not sure this naming is understandable 🤔 What is shipping handling? Can the wording be improved to understand it better?

File: src/Adapter/Carrier/Repository/CarrierRepository.php
Comment (jolelievre) [inline src/Adapter/Carrier/Repository/CarrierRepository.php]: I don't think this is efficient to fetch an object everytime the tax rules need to be updated, because you're supposed to have the object instance already and you'd just need to update it

This is similar to @tleon 's PR https://github.com/PrestaShop/PrestaShop/pull/36246 that is still in progress, but the idea is that `updateInNewVersion` should return the new instance so you can use it afterwards without re-fetching it systematically

File: src/Adapter/Carrier/Validate/CarrierValidator.php
Comment (jolelievre) [inline src/Adapter/Carrier/Validate/CarrierValidator.php:129]: Question, I believe in the UX when free shipping is enabled the other inputs are hidden So it means you can't disable shipping handling once free shipping is enabled

Now the CQRS implementation is not supposed to be influenced by the UX, but it makes me think nonetheless: should this be considered an error or should the `shipping_handling` be disabled automatically when free shipping is enabled? These are two valid approaches so I don't mind either but it's worth thinking about it at least

File: src/Core/Domain/Carrier/Command/AddCarrierCommand.php
Comment (jolelievre) [inline src/Core/Domain/Carrier/Command/AddCarrierCommand.php]:
```suggestion
        private ?string $logoPathName = null
```
Maybe this one could be optional since it's nullable anyway?

File: src/Core/Domain/Carrier/Command/AddCarrierCommand.php
Comment (jolelievre) [inline src/Core/Domain/Carrier/Command/AddCarrierCommand.php]: As commented earlier is there a way to improve this naming? The meaning is not very clear

File: src/Core/Domain/Carrier/Command/AddCarrierCommand.php
Comment (jolelievre) [inline src/Core/Domain/Carrier/Command/AddCarrierCommand.php]:
```suggestion
    public function getTaxRuleGroupId(): int
```

File: tests/Integration/Behaviour/Features/Context/Domain/CarrierFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CarrierFeatureContext.php:114]: I don't know if you knew about it but we have a `PrimitiveUtils` class that helps ith such transformations, although it's optional because the way you used works fine as well Just so you know it has other useful helpers

File: src/Core/Domain/Carrier/Command/EditCarrierCommand.php
Comment (jolelievre) [inline src/Core/Domain/Carrier/Command/EditCarrierCommand.php]: Woooow, this is so not adapted to the usage 😅 But it's already badly design in legacy class:

```php
    /** @var bool Behavior for out-of-range weights: true to disable carrier, false to apply the cost of the highest defined range */
    public $range_behavior;
```

I wonder if we could improve this in the CQRS layer to make it more understandable and self-explanatory. This could be renamed into `useHighestRange` for example, when set to true it uses the highest range, when set to false the carrier is disabled Or maybe `disabledWhenOutOfRange` would be clearer? When set to true you understand that it is disabled and that it's related to the out of range case

The alternative to use the highest one is less explicit and more about a supposedly known convention, but I don't think we can find a way that is clear for both. Unless we completely change the structure and stop using a boolean (in CQRS only), we can use an integer, a string or an enum with two explicit const names: `OutOfRangeBehaviour::USE_HIGHEST_RANGE` and `OutOfRangeBehaviour::DISABLED` then the handler convert them into the legacy boolean

File: src/Adapter/Carrier/CommandHandler/AddCarrierHandler.php
Comment (Hlavtox) [reply src/Adapter/Carrier/CommandHandler/AddCarrierHandler.php]: It's the extra handling fee of the carrier, but I guess the command can now use some better naming right? ~~Also, there should be `get` instead of `is`. Something like `getAdditionalHandlingFee`?~~

File: tests/Integration/Behaviour/Features/Context/Domain/CarrierFeatureContext.php
Comment (boherm) [reply tests/Integration/Behaviour/Features/Context/Domain/CarrierFeatureContext.php:114]: As I see, `PrimitiveUtils::castStringBooleanIntoBoolean` use bootval. But, for some cases like "yes" or "no" for instance, we don't retrieve the correct boolean. So I prefer for the behat tests to have this `filter_var` function. So, in the tests we can use 1, true, yes and  in no case-sensitive way.

File: src/Adapter/Carrier/CommandHandler/AddCarrierHandler.php
Comment (boherm) [reply src/Adapter/Carrier/CommandHandler/AddCarrierHandler.php]: I agree with you guys that the naming are not appropriate here.
But I don't agree with you @Hlavtox about the "should be `get`" because that not the additional fee amount here but the status of if we add additional fees for this carrier.
What about `hasAdditionalHandlingFee`?

File: src/Adapter/Carrier/Validate/CarrierValidator.php
Comment (boherm) [reply src/Adapter/Carrier/Validate/CarrierValidator.php:129]: I like this approach that the `shipping_handling` be disabled automaticaly if `is_free` enabled instead of throwing an exception. I will change that in that way.

File: src/Adapter/Carrier/CommandHandler/AddCarrierHandler.php
Comment (Hlavtox) [reply src/Adapter/Carrier/CommandHandler/AddCarrierHandler.php]: @boherm Sorry, you are right, I am dumb, it's an on/off switch, there should be `IS`. :-)

`hasAdditionalHandlingFee` is good!

File: src/Core/Domain/Carrier/Command/EditCarrierCommand.php
Comment (Hlavtox) [reply src/Core/Domain/Carrier/Command/EditCarrierCommand.php]: @jolelievre 

> Unless we completely change the structure and stop using a boolean (in CQRS only), we can use an integer, a string or an enum with two explicit const names: OutOfRangeBehaviour::USE_HIGHEST_RANGE and OutOfRangeBehaviour::DISABLED then the handler convert them into the legacy boolean

This sounds like the best option, and it also opens the world to another possibilities, like `OutOfRangeBehaviour::SCALE_PROGRESSIVELY`.

File: classes/Carrier.php
Comment (jolelievre) [inline classes/Carrier.php]: I don't think we should change this So far the value is still a `tinyint(1)` so a boolean It's good if we improve this in the CQRS commands/queries but modifying the legacy class doesn't seem relevant

File: src/Adapter/Carrier/CommandHandler/EditCarrierHandler.php
Comment (jolelievre) [inline src/Adapter/Carrier/CommandHandler/EditCarrierHandler.php:94]: Isn't this a bit redundant with the following lines L98-L107, couldn't it be united in the same block for more clarity?

File: classes/Carrier.php
Comment (boherm) [reply classes/Carrier.php]: I agree with you in theory but Phpstan aren't!

Comment (jolelievre) [reply classes/Carrier.php]: I guess it's not happy when you integrate the boolean from the ObjectModel into the QueryResult that uses an integer right? Then you can force casting the boolean into an int, this way you don't need to modify the ObjectModel

File: src/Core/Domain/Carrier/Command/AddCarrierCommand.php
Comment (jolelievre) [inline src/Core/Domain/Carrier/Command/AddCarrierCommand.php]: I think you could introduce a ValueObject for both shippingMehotd and rangeBehavior

Like you did with `OutOfRangeBehaviour`, the value object can validate that the int value is part of a subset, and each allowed value is stored in a const that represents its value like: `OutOfRangeBehaviour::USE_HIGHEST_RANGE`

You can have a similar `ShppingMethod` VO with const values for:
```php
    public const SHIPPING_METHOD_DEFAULT = 0;
    public const SHIPPING_METHOD_WEIGHT = 1;
    public const SHIPPING_METHOD_PRICE = 2;
    public const SHIPPING_METHOD_FREE = 3;
```

Then when you create an add or edit command the input int should be transformed into a VO:
```php
$this->outOfRangerBehaviour = new OutOfRangeBehaviour($rangeBehavior);
```

This will trigger early exceptions with invalid values, which is the workflow used in many other CQRS commands, here's an example:
https://github.com/PrestaShop/PrestaShop/blob/eff832703974bcef13b7ab7ad168a36c7348a48c/src/Core/Domain/Product/Command/UpdateProductCommand.php#L606

This way you don't need to have a public method that is then reused by the validator (like `OutOfRangeBehaviour::assertValue` because the VO validate their content internally

That's the advantage of ValueObject, they are self validating object that prevents using invalid values very early in the process, they will trigger an exception even before the handler or the validator is called, this once you have an instance of a ValueObject you already it can only contain valid value So they are very adapted to this kind of use case with options limited to a subset of values

Comment (jolelievre) [reply src/Core/Domain/Carrier/Command/AddCarrierCommand.php]: Note: our convention is to use VO in commands only, the QueryResult should still return scalar values (so integers in this case)

File: tests/Integration/Behaviour/Features/Scenario/Carrier/carrier_management.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Carrier/carrier_management.feature]: For a more human-friendly test (which is the purpose of gherkin), you should use some strings here that represent the values like in this case: `default` `weight` `price` `free`

It's the behat context step that then transforms the strings into the appropriate integer const values

File: tests/Integration/Behaviour/Features/Scenario/Carrier/carrier_management.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Carrier/carrier_management.feature]: Same suggestion for range behaviour, in the behat features you should use understandable strings

File: tests/Integration/Behaviour/Features/Scenario/Carrier/carrier_management.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Carrier/carrier_management.feature:547]: You tested two use cases when both free shipping and shippingHandling are false initially, which are the to simplest use cases You are missing three alternative scenarios:
- `freeShipping` initially true, `shippingHandling` only is set to true -> `isFree` is automatically turned to `false`
- `shippingHandling` initially true, `freeShipping` only is set to true -> `shippingHandling` is automatically turned to `false`
- when both `shippingHandling` and `isFree` are turned to true an exception is thrown, you need to asert the proper exception with the proper code is thrown

Comment (boherm) [reply tests/Integration/Behaviour/Features/Scenario/Carrier/carrier_management.feature:547]: These tests are already in the PR:
- `Line freeShipping initially true, shippingHandling only is set to true -> isFree is automatically turned to false` -> Start line 372
- `shippingHandling initially true, freeShipping only is set to true -> shippingHandling is automatically turned to false` -> Start line 391
- `when both shippingHandling and isFree are turned to true an exception is thrown, you need to asert the proper exception with the proper code is thrown` -> Start line 410

File: src/Core/Domain/Carrier/Command/EditCarrierCommand.php
Comment (jolelievre) [inline src/Core/Domain/Carrier/Command/EditCarrierCommand.php]:
```suggestion
    public function setAdditionalHandlingFee(bool $hasAdditionalHandlingFee): self
```
It's preferable if getters and setters use the same naming

File: src/Core/Domain/Carrier/ValueObject/ShippingMethod.php
Comment (jolelievre) [inline src/Core/Domain/Carrier/ValueObject/ShippingMethod.php:52]: From the Carrier ObjectModek it seems there are two other possible values for shipping method:
- `SHIPPING_METHOD_FREE` so far I seems this one is used dyanmically when `is_free` is true, but some legacy code still depend on it, maybe we could maintain `shipping_method` automatically updated to this value when `is_free` is turned on?
- `SHIPPING_METHOD_DEFAULT` this seems to avoid defining the shipping method and instead uses the default used one based on the `PS_SHIPPING_METHOD` global configuration, I don't know if it's relevant because it seems to be used for retro compatibility, when we edit a carrier we shouldn't have this option available because you can't define ranges based on a configuration you don't know and that can change

TLDR; maybe these to extra options are not adapted any more and should be ignored in the CQRS layer?

File: src/Core/Form/IdentifiableObject/DataHandler/CarrierFormDataHandler.php
Comment (jolelievre) [inline src/Core/Form/IdentifiableObject/DataHandler/CarrierFormDataHandler.php:61]: The data is ok, but the form will fail once this is merged since the form type has no such data. I guess it's acceptable for now but we must keep that in mind

File: src/Adapter/Carrier/Repository/CarrierRepository.php
Comment (jolelievre) [reply src/Adapter/Carrier/Repository/CarrierRepository.php]: This would be the last big thing to handle, since Thomas's PR is still pending you should probably copy the part from his PR so that we can merge this one sooner

File: classes/Carrier.php
Comment (jolelievre) [reply classes/Carrier.php]: This is still not necessary

File: src/Adapter/Carrier/Validate/CarrierValidator.php
Comment (jolelievre) [inline src/Adapter/Carrier/Validate/CarrierValidator.php]: I think it's worth introducing a specific error code here `INVALID_SHIPPING_HANDLING` is used for invalid types/values But here it's different it's a domain rule that prevents setting to values true at the same time, it should have a dedicated error code so we can also handle a dedicated error message based on it

File: tests/Integration/Behaviour/Features/Scenario/Carrier/carrier_management.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Carrier/carrier_management.feature]: As suggested in another comment, it's worth having a dedicated error code for this domain constraint

**Review by jolelievre — COMMENTED** [review]
Thanks @boherm 

Almost all good, just a few small comments

One big thing left though is about the repository that should return the newly duplicated object model (see Thomas' PR)

File: src/Core/Domain/Carrier/ValueObject/ShippingMethod.php
Comment (boherm) [reply src/Core/Domain/Carrier/ValueObject/ShippingMethod.php:52]: - `SHIPPING_METHOD_FREE` is only when `is_free = true` and `shipping_method` value is set to `SHIPPING_METHOD_FREE` automatically when we make a `getShippingMethod` in Carrier object model. We must always DON'T save this value in database. (think about edit a Carrier with `is_free` off then...)
- `SHIPPING_METHOD_DEFAULT`: As you wish, but it looks like for retro compatibility so....

Comment (Hlavtox) [reply src/Core/Domain/Carrier/ValueObject/ShippingMethod.php:52]: `PS_SHIPPING_METHOD` configuration key is some really old thing that was not even used in 1.6, so it can be safely removed.

I doubt that there is some PS installation where there is a carrier with `SHIPPING_METHOD_DEFAULT` anymore...

Ping @kpodemski

Comment (Hlavtox) [reply src/Core/Domain/Carrier/ValueObject/ShippingMethod.php:52]: Digged deeper, the config is unused since 1.5.5.0, when the carrier wizard was introduced. 👍

https://github.com/PrestaShop/PrestaShop/blob/1.5.5.0/controllers/admin/AdminCarrierWizardController.php

Comment (jolelievre) [reply src/Core/Domain/Carrier/ValueObject/ShippingMethod.php:52]: @boherm then maybe we should clean the `SHIPPING_METHOD_DEFAULT` completely
About the free shipping method, should we handle its value dynamically in the QueryResult as well then?

File: tests/Integration/Behaviour/Features/Scenario/Carrier/carrier_management.feature
Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Scenario/Carrier/carrier_management.feature:547]: My bad I didn't read all correctly, perfect!

File: src/Adapter/Carrier/Repository/CarrierRepository.php
Comment (jolelievre) [reply src/Adapter/Carrier/Repository/CarrierRepository.php]: Still needs to be done, either by copying the code from the pending PR or then this one is blocked until the other is merged

File: src/Core/Domain/Carrier/ValueObject/ShippingMethod.php
Comment (boherm) [reply src/Core/Domain/Carrier/ValueObject/ShippingMethod.php:52]: @jolelievre, we must avoid to do that in the EditableCarrier class, we need here the real value in database.

File: src/Adapter/Carrier/Validate/CarrierValidator.php
Comment (jolelievre) [inline src/Adapter/Carrier/Validate/CarrierValidator.php]:
```suggestion
    private function validateShipping(Carrier $carrier): void
```
Optional, can be done in another PR

**Review by jolelievre — DISMISSED** [review]
Thanks @boherm

**Review by jolelievre — APPROVED** [review]

**Review by tleon — APPROVED** [review]

**Pull request merged**
